keps/sig-node/2902-cpumanager-distribute-cpus-policy-option/README.md

```diff
@@ -0,0 +1,253 @@
+Moreover, these policy options can build on one another, such that multiple tweaks can be made to a policy in an additive fashion.
+The first option introduced in conjunction with #2625 allows one to ensure that only **full** CPUs are allocated to a container, rather than handing out individual hyperthreads from each CPU to different containers.
+This KEP introduces a new `CPUManager` policy option to ensure that CPU allocations are evenly distributed across NUMA nodes in cases where more than one NUMA node is required to satisfy the allocation.
+
+## Motivation
+
+By default, the `CPUManager` will pack CPUs onto one NUMA node until it is filled, with any remaining CPUs simply spilling over to the next NUMA node. 
+This can cause undesired bottlenecks in parallel code relying on barriers (and similar synchronization primitivies), as this type of code tends to run only as fast as its slowest worker (which is slowed down by the fact that fewer CPUs are available on at least one NUMA node).
+By distributing CPUs evenly across NUMA nodes, application developers can more easily ensure that no single worker suffers from NUMA effects more than any other, improving the overall performance of these types of applications.
+
+### Goals
+ * Enable parallel algorithms to run more efficiently when they request more CPUs than can be allocated by a single NUMA node
+
+### Non-Goals
+  * Provide a general solution for all types of CPU distributions across NUMA nodes
+
+## Proposal
+
+We propose to add a new `CPUManager` policy option called `distribute-cpus-across-numa` to the `static` `CPUManager` policy.
+When enabled, this will trigger the `CPUManager` to evenly distribute CPUs across NUMA nodes in cases where more than one NUMA node is required to satisfy the allocation.
+
+### Risks and Mitigations
+
+The risks of adding this new feature are quite low.
+It is isolated to a specific policy option within the `CPUManager`, and is protected both by the option itself, as well as the `CPUManagerPolicyOptions` feature gate (which is disabled by default).
+
+| Risk                                             | Impact | Mitigation |
+| -------------------------------------------------| -------| ---------- |
+| Bugs in the implementation lead to kubelet crash | High   | Disable the policy option and restart the kubelet. The workload will run but with CPU packing semantics - like it was before this new policy option was added. |
+
+## Design Details
+
+When `distribute-cpus-across-numa` is passed as a policy option, the following algorithm will be run to distribute CPUs across NUMA nodes instead of packing them:
+
+```
+Foreach NUMA node:
+   * If all requested CPUs can be allocated from this single NUMA node;
+   --> do the allocation
+
+For each pair of NUMA nodes:
+   * If the set of requested CPUs (modulo 2) can be evenly split across the 2 NUMA nodes; AND
+   * Any remaining CPUs (after the modulo operation) can be striped across some subset of the NUMA nodes;
```

Comment:
In case of remaining CPUs after the modulo operation, how is the subset determined? Is it random or we try to fill the CPUs incrementally based on NUMA Ids?

Reply:
It's the same subset of NUMA node that had CPUs evenly distributed across them. You just now walk through them, one-by-one adding 1 CPU at a time. I guess the choice of _which_ NUMA nodes from this subset will get this extra CPU is somewhat arbitrary. Starting with the lowest number NUMA node and walking towards the highest seems reasonable.

Reply:
Sure, sounds reasonable to me but in cases where we have extra CPUs we could have a decision to make where selection of one NUMA node would favor packing more CPUs on a NUMA whereas another one would result in a balanced allocation of CPUs across NUMA nodes and I was wondering if rather than just simply adding one-by-one it is better to select the one that favors a more balanced allocation of CPUs across NUMA? Nothing changes from workload perspective, but for the subsequent workload we would have more balanced allocation. Here is an example to demonstrate the point:

Considering we have a node with 2 NUMA nodes and CPU distribution like below:

NUMA node 0|  NUMA node 1 
 --- | ---
|cpuID 0  | cpuID 1    
| cpuID 2 | cpuID 3    
| cpuID 4 |  cpuID 5  
| cpuID 6  |  cpuID 7  
|  cpuID 8 | cpuID 9 

With the proposed algorithm, if we have two pods: Pod 1 requesting 1 CPU and Pod2 requesting 5 CPUs, the allocation would look something like below (after performing 5%2 = 1 the remaining 1 CPU comes from NUMA 0 )  :

NUMA node 0|  NUMA node 1 
 --- | ---
|cpuID 0 (allocated to Pod1)  | cpuID 1 (allocated to Pod2)   
| cpuID 2 (allocated to Pod2)| cpuID 3    (allocated to Pod2)
| cpuID 4 (allocated to Pod2)|  cpuID 5  
| cpuID 6 (allocated to Pod2) |  cpuID 7  
|  cpuID 8 | cpuID 9 

So now if a subsequent pod pod 3 requests 4 CPUs comes it means the only option available is to allocate 1 CPU from NUMA node 0 and the other 3 from NUMA node 1, meaning we run into the issue we were trying to solve!? Had we tried to allocate the balanced CPUs such that it was favoring balancing CPUs across NUMA nodes like so 

NUMA  node 0|  NUMA node 1 
--- | ---
|cpuID 0 (allocated to Pod1)  | cpuID 1 (allocated to Pod2)   
| cpuID 2 (allocated to Pod2)| cpuID 3    (allocated to Pod2)
| cpuID 4 (allocated to Pod2)|  cpuID 5  (allocated to Pod2)
| cpuID 6  |  cpuID 7  
|  cpuID 8 | cpuID 9 

 we would be able have a more balanced CPU allocation for Pod 3 like below:

 NUMA node 0|  NUMA node 1 
 --- | ---
|cpuID 0 (allocated to Pod1)  | cpuID 1 (allocated to Pod2)   
| cpuID 2 (allocated to Pod2)| cpuID 3    (allocated to Pod2)
| cpuID 4 (allocated to Pod2)|  cpuID 5  (allocated to Pod2)
| cpuID 6 (allocated to Pod3)  |  cpuID 7  (allocated to Pod3)
|  cpuID 8 (allocated to Pod3) | cpuID 9 (allocated to Pod3)

WDYT?

Reply:
Yeah, I think that makes sense. So instead of just striping arbitrarily, allocate the remaining CPUs to improve the overall balance of CPU allocations across the NUMA nodes (taking all previous allocations into account too).

Reply:
Great, Lets capture this in the KEP.